### PR TITLE
feat(security): client_negotiation handle challenge

### PR DESCRIPTION
**Pull request by levy5307**
client_negotiation handle challenge which is sent from server_negotiation
```
 negotiation process:

                                                  client                             server
                                                     |---   SASL_LIST_MECHANISMS    -->|
                                                     |<-- SASL_LIST_MECHANISMS_RESP ---|
                                                     |---  SASL_SELECT_MECHANISMS  --->|
                                                     |<-- SASL_SELECT_MECHANISMS_OK ---|
                                                     |                                 |
                                                     |------    SASL_INITIATE   ------>|
                                                     |                                 |
                                                     |<----     SASL_CHALLENGE    -----|
                                                     |-----      SASL_RESPONSE    ---->|
                                                     |                                 |
                                                     |              .....              |
                                                     |                                 |
                                                     |<----     SASL_CHALLENGE    -----|
                                                     |-----      SASL_RESPONSE    ---->|
                                                     |                                 |        (authentication will succeed
                                                     |                                 |        if all chanllenges passed)
                                                     | <---        SASL_SUCC      -----|
    (client won't response                           |                                 |
     if servers says ok)                             |                                 |
                                                     | ----         RPC_CALL      ---> |
                                                     | <---         RPC_RESP       ----|
```

### CODE REVIEW

File: src/runtime/security/client_negotiation.cpp [hunk 1]
```diff
@@ -142,6 +142,33 @@ void client_negotiation::on_mechanism_selected(const negotiation_response &resp)
...
+        if (!err.is_ok() && err.code() != ERR_SASL_INCOMPLETE) {
+            dwarn_f("{}: negotiation failed, reason = {}", _name, err.description());
+            fail_negotiation();
+            return;
+        }
+
+        auto req = dsn::make_unique<negotiation_request>();
+        _status = req->status = negotiation_status::type::SASL_CHALLENGE_RESP;
+        req->msg = response_msg;
+        send(std::move(req));
```
Comment (neverchanje) [inline src/runtime/security/client_negotiation.cpp:159]:
```suggestion
        send(negotiation_status::type::SASL_CHALLENGE_RESP, std::move(response_msg));
```

```
void negotiation::send(negotiation_status::type status, std::string && response_msg) {
        auto req = dsn::make_unique<negotiation_request>();
        _status = req->status = negotiation_status::type::SASL_CHALLENGE_RESP;
        req->msg = std::move(response_msg);
        send(std::move(req));
}

void negotiation::send(negotiation_status::type status) {
        auto req = dsn::make_unique<negotiation_request>();
        req->status = status;
        send(std::move(req));
}
```

You can create some overloaded `send` to reduce the repeated code.

**Review by neverchanje — APPROVED** [review]
I've no problem. You can do the tiny refactoring in later PR.

File: src/runtime/security/client_negotiation.cpp (hunk 1 above)
Comment (levy5307) [reply src/runtime/security/client_negotiation.cpp:159]: OK，thanks.

**Review by hycdong — APPROVED** [review]

File: src/runtime/security/client_negotiation.cpp [hunk 2]
```diff
@@ -142,6 +142,33 @@ void client_negotiation::on_mechanism_selected(const negotiation_response &resp)
     }
 }
 
+void client_negotiation::on_challenge(const negotiation_response &challenge)
+{
+    if (challenge.status == negotiation_status::type::SASL_CHALLENGE) {
+        std::string response_msg;
+        auto err = _sasl->step(challenge.msg, response_msg);
+        if (!err.is_ok() && err.code() != ERR_SASL_INCOMPLETE) {
```
Comment (Shuo-Jia) [inline src/runtime/security/client_negotiation.cpp:150]: should `||`？

Comment (levy5307) [reply src/runtime/security/client_negotiation.cpp:150]: No, both `err.is_ok()` and `err.code() == ERR_SASL_INCOMPLETE` means the `_sasl->step()` is ok

**Pull request merged**